### ISSUE TITLE
Create Protocol Buffer (Protobuf) expression capabilities to handle large volumes of data

### ISSUE THREAD

**Issue by dubnemo** [OAGi/Score#832]
**Enhancement request for future roadmap discussion**

In some cases, especially when transferring large arrays of data, a more compressed syntax is required.  Our attempt to replicate a spreadsheet of using JournalEntry resulted in a JSON payload over 1M which is arguably too large for reliable HTTP transmission, requiring compression techniques.  It is a way to send data in multiple related binary files, such as the nested components in OAGIS messages - each core component type would be represented as its own protobuf message type.

Created by Google, Protobuf is being used in many areas.  It has a declarative schema technique, extensible, backward compatibility techniques, and has code generators in multiple languages (C#, Go, Python, Java, etc.).  Current specification is here:
https://developers.google.com/protocol-buffers/docs/proto3

Financial data is a good candidate for Protobuf implementations.  Other areas include Price List, Catalog, and snapshots of manufacturing data and streamed farm operations data within a given start and end time period that need to be exchanged with other parties for analysis.  

This is not intended for transaction based documents such as individual purchase order, shipment, invoice, etc.

**Comment by hakjuoh** [comment]
One drawback protobuf has is that it does not support semantic rules for values such as facets. In order to bring this feature, we may need to adapt third-party plugins like [protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate).

**Comment by dubnemo** [comment]
@hakjuoh it is not clear from reading the semantic rule plug-in you referenced, whether the semantic rule capability is truly needed for serialization of the protobuf definition, or during parsing of protobuf 'instance'.  Is there a way to serialize these semantic rules outside the .proto file?

**Comment by hakjuoh** [comment]
@dubnemo @kbserm As far as I know, there is no way to embed any semantic rule in the protobuf definition. Thus, yes it does if we want to bring it up to the definition. But it is, of course, not "truly" needed as we did for the OpenAPI serialization in Score (e.g. mapping `xsd:gYearMonth` to `type:string` without using a pattern restriction.)
Another concern is to whether define the [services](https://developers.google.com/protocol-buffers/docs/proto3#services), or one might eventually want to adopt it by [gRPC](https://grpc.io/).

**Comment by dubnemo** [comment]
@hakjuoh @kbserm closing this, in lieu of CSV expression which is a more valuable, and common approach.

**Comment by dubnemo** [comment]
@knelson-farmbeltnorth would this issue be of value to ADAPT Standardization?

**Comment by knelson-farmbeltnorth** [comment]
@dubnemo No, not at the currently time.   We standardized on json for anything with a fixed schema.  We use GeoTiff and GeoParquet for high-density spatial data, but these require only a trivial column/attribute list that is defined in the accompanying json.
